### PR TITLE
Measure and output installing and downloading durations

### PR DESCRIPTION
**Pull request by shilovmaksim**
# Description
Improving output for "Install-Binary" and "Start-DownloadWithRetry" functions with additional measuring and logging of installing and downloading durations.

### Test run:
https://github.visualstudio.com/virtual-environments/_build/results?buildId=116599&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=5431112d-2b61-5a5f-7042-ef698f761043&l=5552

## Check list
- [ ] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [x] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 1]
```diff
@@ -49,25 +49,29 @@ function Install-Binary
         $filePath = "msiexec.exe"
     }
 
+    $installStartTime = Get-Date
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:52]: nitpick: move in try-catch block

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 2]
```diff
@@ -49,25 +49,29 @@ function Install-Binary
...
     try
     {
         Write-Host "Starting Install $Name..."
         $process = Start-Process -FilePath $filePath -ArgumentList $ArgumentList -Wait -PassThru
-
         $exitCode = $process.ExitCode
+        $installCompleteTime = [math]::Round(($(Get-Date) - $installStartTime).TotalSeconds, 2)
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:58]: nitpick: Why just don't cast to [int]?

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 3]
```diff
@@ -49,25 +49,29 @@ function Install-Binary
...
         if ($exitCode -eq 0 -or $exitCode -eq 3010)
         {
-            Write-Host "Installation successful"
+            Write-Host "Installation successful in $($installCompleteTime) seconds"
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:61]: nitpick: remove $()

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 4]
```diff
@@ -49,25 +49,29 @@ function Install-Binary
...
         }
         else
         {
             Write-Host "Non zero exit code returned by the installation process: $exitCode"
+            Write-Host "Total time elapsed: $($installCompleteTime) seconds"
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:66]: nitpick: remove $()

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 5]
```diff
@@ -49,25 +49,29 @@ function Install-Binary
...
             exit $exitCode
         }
     }
     catch
     {
+        $installCompleteTime = [math]::Round(($(Get-Date) - $installStartTime).TotalSeconds, 2)
         Write-Host "Failed to install the $fileExtension ${Name}: $($_.Exception.Message)"
+        Write-Host "Installation failed after $($installCompleteTime) seconds"
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:74]: nitpick: remove $()

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 6]
```diff
@@ -179,18 +183,21 @@ function Start-DownloadWithRetry
...
+            $downloadStartTime = Get-Date
             Write-Host "Downloading package from: $Url to path $filePath ."
             (New-Object System.Net.WebClient).DownloadFile($Url, $filePath)
             break
         }
         catch
         {
+            $failTime = [math]::Round(($(Get-Date) - $downloadStartTime).TotalSeconds, 2)
             Write-Host "There is an error during package downloading:`n $_"
+            Write-Host "Total time elapsed $($failTime)"
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:200]: It's not a total time, it's time to a single attempt.

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 [hunk 7]
```diff
@@ -204,6 +211,8 @@ function Start-DownloadWithRetry
         }
     }
 
+    $downloadCompleteTime = [math]::Round(($(Get-Date) - $downloadStartTime).TotalSeconds, 2)
```
Comment (al-cheb) [inline images/win/scripts/ImageHelpers/InstallHelpers.ps1:214]: What about previous attempt if you take the latest one only at 191 line?

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 2 above)
Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1:58]: [int]/[int16]/[int32] so as *.ToInt/*.ToInt16/*.ToInt32 not rounding but taking just whole integer value so might output just 0.
ToDouble decimal .Net method way more complicated.
I wanted to round measured time to second period of milliseconds for better logging and visibility.

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 1 above)
Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Agree. Updated.

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 3 above)
Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Updated.

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 4 above)
Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Updated

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 5 above)
Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Updated

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 7 above)
Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1:216]: Updated with calculation of download attempt time and overall time.

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 6 above)
Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1]: Agree. Missed this point.

File: images/win/scripts/ImageHelpers/InstallHelpers.ps1 (hunk 2 above)
Comment (al-cheb) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1:58]: `[int]/[int16]/[int32] so as .ToInt/.ToInt16/*.ToInt32 not rounding but taking just whole integer value so might output just 0.`
Not exactly, it's called banker's rounding.
```
PS > [int]0.6
1
PS > [int]0.2
0
```

`ToDouble decimal .Net method way more complicated. `
The current type is already Double.

Comment (shilovmaksim) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1:58]: Anyway. Will it be better to round and output like this?
For example for cases like following:
- https://github.visualstudio.com/virtual-environments/_build/results?buildId=116599&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=5431112d-2b61-5a5f-7042-ef698f761043&l=5550
- https://github.visualstudio.com/virtual-environments/_build/results?buildId=116599&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=5431112d-2b61-5a5f-7042-ef698f761043&l=5555

Comment (al-cheb) [reply images/win/scripts/ImageHelpers/InstallHelpers.ps1:58]: LGTM.

**Review by miketimofeev — APPROVED** [review]

**Review by al-cheb — APPROVED** [review]

**Review by dmitry-shibanov — APPROVED** [review]

**Review by vsafonkin — APPROVED** [review]

**Pull request merged**